app/javascript/mastodon/components/hashtag_bar.tsx

```diff
@@ -0,0 +1,195 @@
+import { useState, useCallback } from 'react';
+
+import { FormattedMessage } from 'react-intl';
+
+import { Link } from 'react-router-dom';
+
+import type { List, Record } from 'immutable';
+
+import { groupBy, minBy } from 'lodash';
+
+import { getStatusContent } from './status_content';
+
+// About two lines on desktop
+const VISIBLE_HASHTAGS = 7;
+
+// Those types are not correct, they need to be replaced once this part of the state is typed
+type TagLike = Record<{ name: string }>;
+type StatusLike = Record<{ tags: List<TagLike> }>;
+
+function normalizeHashtag(hashtag: string) {
+  if (hashtag.startsWith('#')) return hashtag.slice(1);
+  else return hashtag;
+}
+
+function isNodeLinkHashtag(element: Node): element is HTMLLinkElement {
+  // it may be a <a> with a hashtag
+  return (
+    element instanceof HTMLAnchorElement &&
+    (element.textContent?.[0] === '#' ||
+      element.previousSibling?.textContent?.[
+        element.previousSibling.textContent.length - 1
+      ] === '#')
+  );
+}
+
+/**
+ * Removes duplicates from an hashtag list, case-insensitive, keeping only the best one
+ * "Best" here is defined by the one with the more casing difference (ie, the most camel-cased one)
+ * @param hashtags The list of hashtags
+ * @returns The input hashtags, but with only 1 occurence of each (case-insensitive)
+ */
+function uniqueHashtagsWithCaseHandling(hashtags: string[]) {
+  const groups = groupBy(hashtags, (tag) => tag.toLowerCase());
```

Comment:
I am using `toLowerCase` here, as using `localeCompare` would be more involved (manually build a list for each new tag, and then compare every tag with every list's keys).

I am not sure this is really worth it, but I can do it if needed.